File: docs/legacy-web/js/finance.js
/**
 * finance.js - Calculs financiers liés au système PV
 *
 * Contient : tarifs EDF, calcul facture, économies, payback actualisé,
 * VAN, LCOE et prime autoconsommation France.
 *
 * Dépendances globales (constants.js) :
 *   ELEC_ESCALATION, PANEL_DEGRADATION, DISCOUNT_RATE, SYSTEM_LIFETIME
 */

const FinanceCalc = (() => {

  // ── Tarifs EDF 2024 (défauts) ──────────────────────────────────
  const TARIFS = {
    base: { price: 0.2516 },
    hphc: { hp: 0.2460, hc: 0.1860 }
  };

  // ── Calcul de la facture annuelle actuelle ─────────────────────
  function calcCurrentAnnualBill(bill) {
    let total = bill.subscriptionPerYear || 0;
    if (bill.tariff === 'base') {
      total += bill.monthlyKwh.reduce((s, k) => s + k * bill.priceBase, 0);
    } else {
      bill.monthlyKwh.forEach((kwh, i) => {
        if (kwh <= 0) return;
        // Fallback 65% HP seulement si aucune donnée HP/HC Enedis disponible.
        // Quand monthlyKwh_hp[i]=0 (toute consommation en HC), ne pas utiliser 0.65.
        const hpRatio = bill.monthlyKwh_hp != null
          ? Math.min(1, Math.max(0, (bill.monthlyKwh_hp[i] ?? 0) / kwh))
          : 0.65;
        const hp = bill.priceHpHc?.hp ?? TARIFS.hphc.hp;
        const hc = bill.priceHpHc?.hc ?? TARIFS.hphc.hc;
        total += kwh * hpRatio * hp + kwh * (1 - hpRatio) * hc;
      });
    }
    return total;
  }

  // ── Calcul des économies selon le tarif ───────────────────────
  function calcSavingsOnBill(monthlyMetrics, bill) {
    if (bill.tariff === 'base') {
      return monthlyMetrics.reduce((sum, m) => sum + m.autoconsoKwh * bill.priceBase, 0);
    }
    // HP/HC : PV produit pendant la journée (heures pleines) → économies au tarif HP
    const hp = bill.priceHpHc?.hp ?? TARIFS.hphc.hp;
    return monthlyMetrics.reduce((sum, m) => sum + m.autoconsoKwh * hp, 0);
  }

  // ── Helpers financiers ────────────────────────────────────────
  // Constantes définies dans constants.js : PANEL_DEGRADATION, ELEC_ESCALATION,
  // DISCOUNT_RATE, SYSTEM_LIFETIME

  /**
   * @param {object|number} [opts]
   * @param {number} [opts.elecEscalation]
   * @param {number} [opts.discountRate]
   * @param {number} [opts.panelDegradation]
   * @param {number} [opts.lifetime]
   */
  function _finOpts(opts) {
    if (typeof opts === 'number') {
      return {
        esc: opts,
        disc: DISCOUNT_RATE,
        deg: PANEL_DEGRADATION,
        life: SYSTEM_LIFETIME
      };
    }
    const o = (opts && typeof opts === 'object') ? opts : {};
    return {
      esc:  (o.elecEscalation   != null && !isNaN(o.elecEscalation))   ? o.elecEscalation   : ELEC_ESCALATION,
      disc: (o.discountRate     != null && !isNaN(o.discountRate))     ? o.discountRate     : DISCOUNT_RATE,
      deg:  (o.panelDegradation != null && !isNaN(o.panelDegradation)) ? o.panelDegradation : PANEL_DEGRADATION,
      life: (o.lifetime != null && !isNaN(o.lifetime) && o.lifetime > 0)
        ? Math.round(o.lifetime) : SYSTEM_LIFETIME
    };
  }

  /**
   * Payback actualisé (DCF, années).
   * Inclut dégradation panneaux, hausse prix électricité, O&M, remplacement onduleur.
   */
  function calcPayback(systemCost, firstYearGain, opts) {
    if (firstYearGain <= 0 || systemCost <= 0) return null;
    const { esc, disc, deg } = _finOpts(opts);
    const omCost      = systemCost * 0.005;
    const inverterRpl = systemCost * 0.12;
    let cum = 0;
    for (let y = 1; y <= 40; y++) {
      const gain = firstYearGain
                 * Math.pow(1 + esc, y - 1)
                 * Math.pow(1 - deg, y - 1);
      const netGain = gain - omCost - (y === 15 ? inverterRpl : 0);
      cum += netGain / Math.pow(1 + disc, y);
      if (cum >= systemCost) return y;
    }
    return null;
  }

  /**
   * Valeur Actuelle Nette (€) sur l'horizon choisi.
   * VAN > 0 → investissement rentable au taux d'actualisation.
   */
  function calcNPV(systemCost, firstYearGain, opts) {
    if (systemCost <= 0) return 0;
    if (firstYearGain <= 0) return -systemCost;
    const { esc, disc, deg, life } = _finOpts(opts);
    const omCost      = systemCost * 0.005;
    const inverterRpl = systemCost * 0.12;
    let npv = -systemCost;
    for (let y = 1; y <= life; y++) {
      const gain = firstYearGain
                 * Math.pow(1 + esc, y - 1)
                 * Math.pow(1 - deg, y - 1);
      const netGain = gain - omCost - (y === 15 ? inverterRpl : 0);
      npv += netGain / Math.pow(1 + disc, y);
    }
    return npv;
  }

  /**
   * LCOE (€/kWh) avec dégradation + maintenance annuelle + remplacement onduleur.
   */
  function calcLCOE(systemCost, annualProd, opts) {
    if (annualProd <= 0 || systemCost <= 0) return 0;
    const { deg, life } = _finOpts(opts);
    const omRate       = 0.005; // 0.5 %/an O&M
    const inverterRepl = systemCost * 0.12;
    let cumProd = 0, cumCost = systemCost;
    for (let y = 1; y <= life; y++) {
      cumProd += annualProd * Math.pow(1 - deg, y - 1);
      cumCost += systemCost * omRate;
      if (y === 15) cumCost += inverterRepl;
    }
    return cumCost / cumProd;
  }

  /**
   * Prime à l'autoconsommation solaire (France - décret 2021-1444).
   * Varie chaque trimestre - vérifier l'arrêté en vigueur sur energie.gouv.fr
   * Valeurs indicatives 2025 (à vérifier sur energie.gouv.fr chaque trimestre) :
   *   ≤ 3 kWc  : 300 €/kWc  |  ≤ 9 kWc : 230 €/kWc
   *   ≤ 36 kWc : 100 €/kWc  |  ≤ 100 kWc : 60 €/kWc
   */
  function calcFrenchIncentive(Ppeak) {
    if (Ppeak <= 0)   return 0;
    if (Ppeak <= 3)   return Math.round(Ppeak * 300);
    if (Ppeak <= 9)   return Math.round(Ppeak * 230);
    if (Ppeak <= 36)  return Math.round(Ppeak * 100);
    if (Ppeak <= 100) return Math.round(Ppeak * 60);
    return 0;
  }

  return { TARIFS, calcCurrentAnnualBill, calcSavingsOnBill, calcPayback, calcNPV, calcLCOE, calcFrenchIncentive };
})();

File: src/core/finance.cpp
#include "finance.h"

#include "constants.h"

#include <cmath>

namespace ose {

Finance::Finance(QObject* parent) : QObject(parent) {}

namespace {

struct FinOpts {
    double esc = kElecEscalation;
    double disc = kDiscountRate;
    double deg = kPanelDegradation;
    int life = kSystemLifetime;
};

FinOpts finOpts(const QVariantMap& opts)
{
    FinOpts o;
    if (opts.contains(QStringLiteral("elecEscalation")))
        o.esc = opts.value(QStringLiteral("elecEscalation")).toDouble();
    if (opts.contains(QStringLiteral("discountRate")))
        o.disc = opts.value(QStringLiteral("discountRate")).toDouble();
    if (opts.contains(QStringLiteral("panelDegradation")))
        o.deg = opts.value(QStringLiteral("panelDegradation")).toDouble();
    if (opts.contains(QStringLiteral("lifetime")))
        o.life = opts.value(QStringLiteral("lifetime")).toInt();
    return o;
}

} // namespace

double Finance::calcCurrentAnnualBill(const QVariantMap& bill) const
{
    double total = bill.value(QStringLiteral("subscriptionPerYear")).toDouble();
    const QVariantList monthly = bill.value(QStringLiteral("monthlyKwh")).toList();
    const QString tariff = bill.value(QStringLiteral("tariff"), QStringLiteral("base")).toString();
    if (tariff == QLatin1String("base")) {
        const double price = bill.value(QStringLiteral("priceBase"), 0.2516).toDouble();
        for (const QVariant& k : monthly)
            total += k.toDouble() * price;
    } else {
        const QVariantList hpList = bill.value(QStringLiteral("monthlyKwh_hp")).toList();
        const QVariantMap priceHpHc = bill.value(QStringLiteral("priceHpHc")).toMap();
        const double hp = priceHpHc.value(QStringLiteral("hp"), 0.2460).toDouble();
        const double hc = priceHpHc.value(QStringLiteral("hc"), 0.1860).toDouble();
        for (int i = 0; i < monthly.size(); ++i) {
            const double kwh = monthly[i].toDouble();
            if (kwh <= 0)
                continue;
            double hpRatio = 0.65;
            if (!hpList.isEmpty() && i < hpList.size())
                hpRatio = std::min(1.0, std::max(0.0, hpList[i].toDouble() / kwh));
            total += kwh * hpRatio * hp + kwh * (1 - hpRatio) * hc;
        }
    }
    return total;
}

double Finance::calcSavingsOnBill(const QVariantList& monthlyMetrics, const QVariantMap& bill) const
{
    if (bill.value(QStringLiteral("tariff")).toString() == QLatin1String("base")) {
        const double price = bill.value(QStringLiteral("priceBase"), 0.2516).toDouble();
        double sum = 0;
        for (const QVariant& v : monthlyMetrics)
            sum += v.toMap().value(QStringLiteral("autoconsoKwh")).toDouble() * price;
        return sum;
    }
    const QVariantMap priceHpHc = bill.value(QStringLiteral("priceHpHc")).toMap();
    const double hp = priceHpHc.value(QStringLiteral("hp"), 0.2460).toDouble();
    double sum = 0;
    for (const QVariant& v : monthlyMetrics)
        sum += v.toMap().value(QStringLiteral("autoconsoKwh")).toDouble() * hp;
    return sum;
}

QVariant Finance::calcPayback(double systemCost, double firstYearGain, const QVariantMap& opts) const
{
    if (firstYearGain <= 0 || systemCost <= 0)
        return QVariant();
    const FinOpts o = finOpts(opts);
    const double omCost = systemCost * 0.005;
    const double inverterRpl = systemCost * 0.12;
    double cum = 0;
    for (int y = 1; y <= 40; ++y) {
        const double gain = firstYearGain * std::pow(1 + o.esc, y - 1) * std::pow(1 - o.deg, y - 1);
        const double netGain = gain - omCost - (y == 15 ? inverterRpl : 0);
        cum += netGain / std::pow(1 + o.disc, y);
        if (cum >= systemCost)
            return y;
    }
    return QVariant();
}

double Finance::calcNPV(double systemCost, double firstYearGain, const QVariantMap& opts) const
{
    if (systemCost <= 0)
        return 0;
    if (firstYearGain <= 0)
        return -systemCost;
    const FinOpts o = finOpts(opts);
    const double omCost = systemCost * 0.005;
    const double inverterRpl = systemCost * 0.12;
    double npv = -systemCost;
    for (int y = 1; y <= o.life; ++y) {
        const double gain = firstYearGain * std::pow(1 + o.esc, y - 1) * std::pow(1 - o.deg, y - 1);
        const double netGain = gain - omCost - (y == 15 ? inverterRpl : 0);
        npv += netGain / std::pow(1 + o.disc, y);
    }
    return npv;
}

double Finance::calcLCOE(double systemCost, double annualProd, const QVariantMap& opts) const
{
    if (annualProd <= 0 || systemCost <= 0)
        return 0;
    const FinOpts o = finOpts(opts);
    constexpr double omRate = 0.005;
    const double inverterRepl = systemCost * 0.12;
    double cumProd = 0;
    double cumCost = systemCost;
    for (int y = 1; y <= o.life; ++y) {
        cumProd += annualProd * std::pow(1 - o.deg, y - 1);
        cumCost += systemCost * omRate;
        if (y == 15)
            cumCost += inverterRepl;
    }
    return cumCost / cumProd;
}

double Finance::calcFrenchIncentive(double Ppeak) const
{
    if (Ppeak <= 0)
        return 0;
    if (Ppeak <= 3)
        return std::round(Ppeak * 300);
    if (Ppeak <= 9)
        return std::round(Ppeak * 230);
    if (Ppeak <= 36)
        return std::round(Ppeak * 100);
    if (Ppeak <= 100)
        return std::round(Ppeak * 60);
    return 0;
}

} // namespace ose
